tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,376 @@
+
+        """
+
+        def conv_act_branch(
+            layer: torch.nn.Module, in_tensor: torch.Tensor
+        ) -> torch.Tensor:
+            """Applies Convolution and Activation to the input tensor.
+
+            Args:
+                layer (torch.nn.ModuleDict):
+                    Torch layer as ModuleDict.
+                in_tensor (torch.Tensor):
+                    Input Tensor.
+            Returns:
+                torch.Tensor:
+                    Torch Tensor after applying convolution and activation.
+
+            """
+            return layer["conv1"](in_tensor)
+
+        def spatially_constrained_layer1(
+            layer: torch.nn.Module,
+            in_tensor: torch.Tensor,
+            out_height: int = 13,
+            out_width: int = 13,
+        ) -> tuple:
+            """Spatially constrained layer 1.
+
+            Estimates row, column and height for
+            spatially_constrained_layer2 layer mapping.
+
+            Args:
+                layer (torch.nn.Module):
+                    Torch layer as ModuleDict.
+                in_tensor (torch.Tensor):
+                    Input Tensor.
+                out_height (int):
+                    Output height.
+                out_width (int):
+                    Output Width
+
+            Returns:
+                tuple of :class:`torch.Tensor`:
+                    Row, Column and height estimates used for
+                    spatially_constrained_layer2 mapping.
+
+            """
+            sigmoid = layer["conv1"](in_tensor)
+            sigmoid0 = sigmoid[:, 0:1, :, :] * (out_height - 1)
+            sigmoid1 = sigmoid[:, 1:2, :, :] * (out_width - 1)
+            sigmoid2 = sigmoid[:, 2:3, :, :]
+            return sigmoid0, sigmoid1, sigmoid2
+
+        def spatially_constrained_layer2(network, sc1_0, sc1_1, sc1_2) -> torch.Tensor:
+            """Spatially constrained layer 2.
+
+            Estimates row, column and height for sc2 layer mapping.
+
+            Args:
+                network (:class:`.SCCNN`):
+                    An initiated SCCNN class.
+                sc1_0 (torch.Tensor):
+                    Output of spatially_constrained_layer1 estimating
+                    the x position of the nucleus.
+                sc1_1 (int):
+                    Output of spatially_constrained_layer1 estimating
+                    the y position of the nucleus.
+                sc1_2 (int):
+                    Output of spatially_constrained_layer1 estimating
+                    the confidence in nucleus detection.
+
+            Returns:
+                :class:`torch.Tensor`:
+                    Probability map using the estimates from
+                    spatially_constrained_layer1.
+
+            """
+            x = torch.tile(
+                network.x, dims=[sc1_0.size(0), 1, 1, 1]
+            )  # Tile for batch size
+            y = torch.tile(network.y, dims=[sc1_0.size(0), 1, 1, 1])
```

Comment:
`self.x` and `self.y` are not properly defined variables. If you want the constant variables to go together with all the weights, consider using `self.register_buffer` in init

```
        self.register_buffer(
            "unpool_mat", torch.from_numpy(np.ones((2, 2), dtype="float32"))
        )
```

else `model.to(device)` may have unintended behavior as `self.x` and `self.y` are defined only on the main device.

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/434/commits/b24422f2f1d43f39844ef603bb68db0a9bcc2757